src/webui/webapplication.cpp

```diff
@@ -173,45 +98,123 @@ namespace
             return TriStateBool::False;
         return TriStateBool::Undefined;
     }
+
+    void translateDocument(QString &data)
+    {
+        const QRegExp regex("QBT_TR\\((([^\\)]|\\)(?!QBT_TR))+)\\)QBT_TR(\\[CONTEXT=([a-zA-Z_][a-zA-Z0-9_]*)\\])");
+        const QRegExp mnemonic("\\(?&([a-zA-Z]?\\))?");
+        int i = 0;
+        bool found = true;
+
+        const QString locale = Preferences::instance()->getLocale();
+        bool isTranslationNeeded = !locale.startsWith("en") || locale.startsWith("en_AU") || locale.startsWith("en_GB");
+
+        while (i < data.size() && found) {
+            i = regex.indexIn(data, i);
+            if (i >= 0) {
+                //qDebug("Found translatable string: %s", regex.cap(1).toUtf8().data());
+                QByteArray word = regex.cap(1).toUtf8();
+
+                QString translation = word;
+                if (isTranslationNeeded) {
+                    QString context = regex.cap(4);
+                    translation = qApp->translate(context.toUtf8().constData(), word.constData(), 0, 1);
+                }
+                // Remove keyboard shortcuts
+                translation.replace(mnemonic, "");
+
+                // Use HTML code for quotes to prevent issues with JS
+                translation.replace("'", "&#39;");
+                translation.replace("\"", "&#34;");
+
+                data.replace(i, regex.matchedLength(), translation);
+                i += translation.length();
+            }
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    for (int i = metaObject()->methodOffset(); i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (QLatin1String(method.tag()) == QLatin1String("WEBAPI_PUBLIC"))
+            m_publicAPIs << method.name();
+        else if (QLatin1String(method.tag()) == QLatin1String("WEBAPI_PRIVATE"))
+            m_privateAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
+        status(404, "Not Found");
```

Comment:
is `500 "Internal Server Error"` more suitable here?

Reply:
>The 500 (Internal Server Error) status code indicates that the server
   encountered an unexpected condition that prevented it from fulfilling
   the request.

We hasn't "unexpected condition" here.
In fact, the HTTP has not appropriate codes for some of our situations.
I imply once we do not support paths with `.` and `..` in it, we have no resources with these paths, so any such resources are "Not Found".

Reply:
The another situation is what we should return, for example, when user request the creation of something?
1. the resource with same id is already exists
2. unsupported id or another data

Reply:
>I imply once we do not support paths with . and .. in it, we have no resources with these paths, so any such resources are "Not Found".

My concern is returning 404 when the resource is indeed there is confusing (e.g. `./image.jpg`, web servers like apache, nginx does not have this limit), so IMO 500 (or 403) is more explicit to the user that we don't accept `.` or `..`.
I'll let you decide this.

>the resource with same id is already exists

Depends on the context... 200 OK (if we are following RESTful style), or 409

> unsupported id or another data

Hard to find an answer without context here, 404, 403?

Reply:
>The another situation is what we should return, for example, when user request the creation of something?
1. the resource with same id is already exists
2. unsupported id or another data

@Chocobo1, I mean, e.g., addCategory: 1. when category with the given name is already exists; 2. when request has unacceptable category name (e.g. empty).

Can you also revise all statuses returned from WebApplication and give your suggestions.

Reply:
>Can you also revise all statuses returned from WebApplication and give your suggestions.

`409 Conflict`: increasePrio(), decreasePrio(), topPrio(), bottomPrio(), the first error in rename(), setCategory(), addCategory()

Also please change to use the standard status name e.g.`Not Found` in `status()` and use `print()` to embed the specific error.

Reply:
Done.